packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -446,5 +452,675 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
+        // accommodate the back button.
+      }
+    } else {
+      // No divider for the next button when there's only one page.
+      parentWidth -= dividerWidth;
+    }
+
+    size = Size(parentWidth, _kToolbarHeight);
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    visitChildren((RenderObject renderObjectChild) {
+      final RenderBox child = renderObjectChild as RenderBox;
+      final ToolbarItemsParentData childParentData = child.parentData as ToolbarItemsParentData;
+
+      if (childParentData.shouldPaint) {
+        final Offset childOffset = childParentData.offset + offset;
+        context.paintChild(child, childOffset);
+      }
+    });
+  }
+
+  @override
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! ToolbarItemsParentData) {
+      child.parentData = ToolbarItemsParentData();
+    }
+  }
+
+  // Returns true iff the single child is hit by the given position.
+  static bool hitTestChild(RenderBox child, BoxHitTestResult result, { Offset position }) {
+    if (child == null) {
+      return false;
+    }
+    final ToolbarItemsParentData childParentData =
+        child.parentData as ToolbarItemsParentData;
+    return result.addWithPaintOffset(
+      offset: childParentData.offset,
+      position: position,
+      hitTest: (BoxHitTestResult result, Offset transformed) {
+        assert(transformed == position - childParentData.offset);
+        return child.hitTest(result, position: transformed);
+      },
+    );
+  }
+
+  @override
+  bool hitTestChildren(BoxHitTestResult result, { Offset position }) {
+    // Hit test list children.
+    // The x, y parameters have the top left of the node's box as the origin.
+    RenderBox child = lastChild;
+    while (child != null) {
+      final ToolbarItemsParentData childParentData = child.parentData as ToolbarItemsParentData;
+
+      // Don't hit test children that aren't shown.
+      if (!childParentData.shouldPaint) {
+        child = childParentData.previousSibling;
+        continue;
+      }
+
+      if (hitTestChild(child, result, position: position)) {
+        return true;
+      }
+      child = childParentData.previousSibling;
+    }
+
+    // Hit test slot children.
+    if (hitTestChild(backButton, result, position: position)) {
+      return true;
+    }
+    if (hitTestChild(nextButton, result, position: position)) {
+      return true;
+    }
+    if (hitTestChild(nextButtonDisabled, result, position: position)) {
+      return true;
+    }
+
+    return false;
+  }
+
+  @override
+  void attach(PipelineOwner owner) {
+    super.attach(owner);
+    visitChildren((RenderObject renderObjectChild) {
+      final RenderBox child = renderObjectChild as RenderBox;
+      child.attach(owner);
+    });
+  }
+
+  @override
+  void detach() {
+    // Detach list children.
+    super.detach();
+
+    // Detach slot children.
+    visitChildren((RenderObject renderObjectChild) {
```

Comment:
Instead of iterating over all children, couldn't you just iterate over the slot children in childToSlot?